hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/upgrade/UpgradeTask.java

```diff
@@ -0,0 +1,391 @@
+      return result;
+    });
+  }
+
+  private List<Result> upgradeSubContainerDir(File containerTopDir)
+      throws IOException {
+    List<Result> resultList = new ArrayList<>();
+    if (containerTopDir.isDirectory()) {
+      File[] containerDirs = containerTopDir.listFiles();
+      if (containerDirs != null) {
+        for (File containerDir : containerDirs) {
+          final ContainerData containerData = parseContainerData(containerDir);
+          if (containerData != null) {
+            final Result result = new Result(containerData);
+            upgradeContainer(containerData, result);
+            resultList.add(result);
+          }
+        }
+      }
+    }
+    return resultList;
+  }
+
+  private ContainerData parseContainerData(File containerDir) {
+    try {
+      File containerFile = ContainerUtils.getContainerFile(
+          containerDir);
+      long containerID =
+          ContainerUtils.getContainerID(containerDir);
+      if (containerFile.exists()) {
+        try {
+          ContainerData containerData =
+              ContainerDataYaml.readContainerFile(
+                  containerFile);
+          if (containerID != containerData.getContainerID()) {
+            LOG.error("Invalid ContainerID in file {}. " +
+                    "Skipping loading of this container.",
+                containerFile);
+            return null;
+          }
+          if (containerData.getContainerType().equals(
+              ContainerProtos.ContainerType.KeyValueContainer) &&
+              containerData instanceof KeyValueContainerData) {
+            KeyValueContainerData kvContainerData =
+                (KeyValueContainerData)
+                    containerData;
+            containerData.setVolume(hddsVolume);
+            KeyValueContainerUtil
+                .parseKVContainerData(kvContainerData, config);
+
+            return kvContainerData;
+          } else {
+            LOG.error("Only KeyValueContainer support. ");
+            return null;
+          }
+
+        } catch (IOException ex) {
+          LOG.error(
+              "Failed to parse ContainerFile for ContainerID: {}",
+              containerID, ex);
+          return null;
+        }
+      } else {
+        LOG.error("Missing .container file for ContainerID: {}",
+            containerDir.getName());
+        return null;
+      }
+    } catch (Throwable e) {
+      LOG.error("Failed to load container from {}",
+          containerDir.getAbsolutePath(), e);
+      return null;
+    }
+  }
+
+  private void upgradeContainer(ContainerData containerData, Result result)
+      throws IOException {
+    final DBStore targetDBStore = datanodeStoreSchemaThree.getStore();
+
+    // open container schema v2 rocksdb
+    final DatanodeStore dbStore = BlockUtils
+        .getUncachedDatanodeStore((KeyValueContainerData) containerData, config,
+            true);
+    final DBStore sourceDBStore = dbStore.getStore();
+
+    final String blockDataTableName =
```

Comment:
Can we call DatanodeSchemaTwoDBDefinition.getMap() and iterate all the items to do the transfor?